### DISCUSSION TITLE
CMT Frequenzänderung über web api mit http-post funktioniert nicht

### DISCUSSION THREAD

**Discussion by khschmidt** [tbnobody/OpenDTU#2467] [Q&A] [answered]
### What happened?

Ich lese über python das json Objekt über http-get von /api/dtu/config aus, ändere dann die Frequenz unter cmt_frequency und sende das vollständige json Objekt über http-post an die DTU. Ich erhalte dann eine Fehlermeldung: Failed to parse data. Wenn ich mit derselben Vorgangsweise /api/ntp/config einlese und dann ntp_server ändere und dann mit http-post an die DTU sende, dann wird die Änderung übernommen. Prinzipiell sollte meine Vorgangsweise also korrekt sein.

### To Reproduce Bug

#python script:
import requests, time, sys, logging, pushover
from requests.auth import HTTPBasicAuth

serial = "1164a006b111" # Seriennummer des Hoymiles Wechselrichters, Buchstaben klein!!!

dtu_ip = '10.0.0.206' # IP Adresse von OpenDTU
dtu_nutzer = 'admin' # OpenDTU Nutzername
dtu_passwort = 'blablabla' # OpenDTU Passwort

dtu = requests.get(url = f'http://{dtu_ip}/api/dtu/config', auth = HTTPBasicAuth(dtu_nutzer, dtu_passwort) ).json()
freq = int(dtu['cmt_frequency'])
dtu['cmt_frequency'] = freq + 250000
print(dtu)
try:
    r = requests.post(
    url = f'http://{dtu_ip}/api/dtu/config',
    data = f'data={dtu}',
    auth = HTTPBasicAuth(dtu_nutzer, dtu_passwort),
    headers = {'Content-Type': 'application/x-www-form-urlencoded'}
    )
    print(f'Konfiguration gesendet ({r.json()})')
    #print(r)
except:
    print('Fehler beim Senden der Konfiguration')
dtu = requests.get(url = f'http://{dtu_ip}/api/dtu/config', auth = HTTPBasicAuth(dtu_nutzer, dtu_passwort) ).json()
freq = int(dtu['cmt_frequency'])
print(freq)

### Expected Behavior

wie beschrieben möchte ich cmt_frequency über die web api ändern

### Install Method

Pre-Compiled binary from GitHub releases

### What git-hash/version of OpenDTU?

24.9.30

### What firmware variant (PIO Environment) are you using?

generic_esp32

### Relevant log/trace output

```Shell
json Antwort von http-post:
Konfiguration gesendet ({'type': 'warning', 'message': 'Failed to parse data!', 'code': 1004})
```

### Please confirm the following

- [X] I believe this issue is a bug that affects all users of OpenDTU, not something specific to my installation.
- [X] I have already searched for relevant existing issues and discussions before opening this report.
- [X] I have updated the title field above with a concise description.
- [X] I have double checked that my inverter does not contain a W in the model name (like HMS-xxxW) as they are not supported.

**Comment by tbnobody** [comment]
Moving it to discussions as it is no issue but most likely a faulty usage of the API. (The web frontend only uses the API you described and its working there)

**Comment by tbnobody** [comment, marked as answer]
The error message `Failed to parse data!', ` indicates that your json is invalid.

See https://github.com/tbnobody/OpenDTU/blob/3dc70ab40aade8b7eb9ed9a1c6605ca326299d18/src/WebApi.cpp#L108-L115

**Comment by khschmidt** [comment]
Ist der Inhalt des json-Objekts bei /api/dtu/config unterschiedlich aufgebaut bei HTTP-get und HTTP-post?

**Reply by stefan123t** [reply]
Du gibst im zweiten request gar kein json() mit sondern verwendest vermutlich url_parameter bzw. form data. Der Aufruf ist mE nicht korrekt:

```python
r = requests.post(
url = f'http://{dtu_ip}/api/dtu/config',
data = f'data={dtu}',
auth = HTTPBasicAuth(dtu_nutzer, dtu_passwort),
headers = {'Content-Type': 'application/x-www-form-urlencoded'}
)
```

Trace doch die Änderung mal im Browser mit den Developer Tools (F12) > Network mit und schau Dir u.a. auch die verwendeten Headers an.

**Comment by khschmidt** [comment]
@stefan123t danke für den Hinweis mit den Developertools, habe ich zwar schon gekannt, aber in diesem Fall nicht daran gedacht. Der python code ist korrekt und funktioniert bei Limitänderung seit Monaten problemlos. Auch die Änderung des NTP-Servers funktionierte sofort. Nach mehreren Stunden erfolglosen Suchens nach einem Fehler in meiner request.post Anweisung ist mir dann eingefallen, dass ich bei den ersten Versuchen bei der Limit-Vorgabe ein ähnliche Problem hatte, damals hatte ich Großbuchstaben in der Seriennummer und das hat nicht funktioniert. Nun sind in dem JSON-Objekt /api/dtu/config Werte, die TRUE oder FALSE sein können. Die json Lib von python übersetzt dies mit True und False. Erlaubt sind aber offensichtlich nur true und false.
Korrigiert man diese Werte im JSON-Objekt auf Kleinschreibung dann funktioniert die Änderung der cmt_frequency sofort.
Problem gelöst: Konfiguration gesendet ({'type': 'success', 'message': 'Settings saved!', 'code': 1001})
